File: ts/src/txt2sql.ts
import {promises as fs} from 'fs';
import * as Path from 'path';

export type Txt2SqlOptions = {
    field_separator: '' | '|' | ',' | ';' | '\t' | '...'
    table_name?: string
    quote_identifiers: string
    case: 'lower' | 'upper' | 'mixed'
    insert_columns: boolean
    multi_insert: boolean
    infer_types: boolean
    output_types: boolean
  }

export const txt2sqlOption = {
    for_test: { 
        field_separator: '',
        insert_columns: false,
        multi_insert: true,
        quote_identifiers: '',
        case: 'mixed',
        infer_types: false,
        output_types: false
    } as Txt2SqlOptions
}

export interface OutWritter {
    write(paragraph:string):Promise<void>
    close():Promise<void> // no more writes/closes allowed
}

export interface OutWritterFactory {
    open(fileName:string):Promise<OutWritter>
    end():Promise<void> // no more writes/closes allowed in any opened file
}

export class FileWritter implements OutWritter{
    constructor(public file:fs.FileHandle){}
    async write(paragraph:string){
        await this.file.write(paragraph);
    }
    async close(){
        return this.file.close()
    }
}

export class FileWritterFactory implements OutWritterFactory{
    public extension = '.sql'
    async open(fileName:string){
        if(!Path.extname(fileName)){
            fileName += this.extension;
        }
        return new FileWritter(await fs.open(fileName,undefined,'w'));
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    async end(){}
}

abstract class TypeBase<T> {
    abstract parseValue(rawValue:string):T|null
    abstract quoteValue(value:T):string
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    registerOptions(_value:T, _rawValue:string){}
}

class TypeInferer<T, T2 extends TypeBase<T>>{
    maxValue:T|null = null
    minValue:T|null = null
    otherValues = false
    hasNull = false
    hasInvalid:boolean = false
    constructor(public typer:T2){}
    inferValue(rawValue:string|null){
        if(this.hasInvalid) return;
        if(rawValue == null){
            this.hasNull = true;
        } else {
            const value = this.typer.parseValue(rawValue);
            if(value == null){ 
                this.hasInvalid = true;
                this.hasInvalid = rawValue as unknown as true;
                return;
            }
            this.typer.registerOptions(value, rawValue);
            if(this.maxValue == undefined || this.minValue == undefined ){
                this.maxValue = value
                this.minValue = value
            }else if(value >= this.maxValue){
                this.maxValue = value
            } else if(value <= this.minValue){
                this.minValue = value
            } else {
                this.otherValues = true;
            }
        }
    }
}

class TypeProcessor<T, T2 extends TypeBase<T>>{
    constructor(protected typer:T2, protected columnName:string){}
    processValue(rawValue:string|null, line:number):string{
        if(rawValue == null){
            return 'null';
        }
        const value = this.typer.parseValue(rawValue);
        if(value == undefined){ 
            throw new Error(`invalid value "${rawValue}" in column "${this.columnName}" line ${line}`)
        }
        return this.typer.quoteValue(value);
    }
}

export type Constructor<T> = new(...args: unknown[]) => T;
export type AbstractConstructor<T> = abstract new(...args: unknown[]) => T;

type TypeConstructor = typeof TypeBase
// type TypeConstructor = Constructor<TypeBase<unknown>>

class StringType extends TypeBase<string>{
    maxLength = 0
    quoteValue(value: string): string {
        return "'" + value.replace(/'/g, "''") + "'";
    }
    parseValue(rawValue: string): string {
        return rawValue;
    }
    override registerOptions(value: string): void {
        if(value.length > this.maxLength) this.maxLength = value.length;
    }
}

class NumberType extends TypeBase<number>{
    protected hasDecimal = false
    protected rawDecimal = '.'
    quoteValue(value: number): string {
        return value.toString();
    }
    parseValue(rawValue: string): number | null {
        const value = Number(rawValue);
        if(isNaN(value)){
            return null
        }
        return value
    }
    override registerOptions(_value: number, rawValue:string): void {
        if(rawValue.includes(this.rawDecimal)){
            this.hasDecimal = true;
        }
    }
}

const numberSignInverter = {'.':',', ',':'.'};

class SpanishNumberType extends NumberType{
    constructor(){
        super()
        this.rawDecimal = ','
    }
    override parseValue(rawValue: string): number | null {
        return super.parseValue(rawValue.replace(/[.,]/g, (sign:string) => numberSignInverter[sign as '.'|',']))
    }
}

type ColumnInfoTypeRaw = {type_name: 'text', max_length: null}

type ColumnInfoTypes = ColumnInfoTypeRaw

type ColumnInfo = ColumnInfoTypes & {name:string}

var rawColumninfo: ColumnInfoTypeRaw = {type_name:'text', max_length: null}

enum Part { Body, Head }

type Txt2SqlOpenedMembers = {
    columnProcessor: TypeProcessor<unknown, TypeBase<unknown>>[]
    columnInfo: ColumnInfo[]
    create_table_raw: OutWritter
    inserts: OutWritter
    create_table?: OutWritter
    part: Part
    rowNumber: number
}

enum Status { New, Started, Done }

type Txt2SqlStatus = { is: Status.New | Status.Done } | { is: Status.Started } & Txt2SqlOpenedMembers

function createNew<T extends TypeBase<unknown>>(c:AbstractConstructor<T>):T{
    //@ts-expect-error For create without estends
    return new c();
}

export class Txt2Sql{
    private status: Txt2SqlStatus = { is: Status.New }
    private table_name;
    protected columnInfering: TypeInferer<unknown, TypeBase<unknown>>[][] = []
    constructor(private owf: OutWritterFactory, public options:Txt2SqlOptions = txt2sqlOption.for_test){
        this.table_name = this.getIdentifier(this.options.table_name || 'table1');
    }
    caseIdentifier(identifier:string){
        return this.options.case == 'mixed' ? identifier : 
            (identifier[this.options.case == 'upper' ? 'toLocaleUpperCase' : 'toLocaleLowerCase']())
    }
    quoteIdentifier(identifier:string):string{
        const q = this.options.quote_identifiers;
        const replaced = q != '' ? q + q : '_';
        return q + identifier.replace(q != '' ? new RegExp(q, 'g') : /\W/g, replaced) + q;
    }
    getIdentifier(identifier:string){
        return this.quoteIdentifier(this.caseIdentifier(identifier))
    }
    async processStart(){
        switch(this.status.is){
            case Status.New:
                this.status = {
                    is: Status.Started,
                    columnInfo: [],
                    columnProcessor: [],
                    create_table_raw: await this.owf.open('create_table_raw'),
                    inserts: await this.owf.open('inserts'),
                    part: Part.Head,
                    rowNumber: 0
                }
                if(this.options.infer_types){
                    this.columnInfering = []
                }
            break;
            default: throw Error("wrong status at process_end: "+this.status);
        }
    }
    inferType(rawValue:string, columnNumber:number){
        if(this.status.is != Status.Started) throw new Error('cannot inferType if not Started')
        const columnInfering = this.columnInfering[columnNumber];
        columnInfering.forEach(inferer => inferer.inferValue(rawValue));
    }
    typeConstructors():TypeConstructor[]{
        return [NumberType, SpanishNumberType, StringType];
    }
    processColumns(rawNames:string[]){
        if(this.status.is != Status.Started) throw new Error('cannot processColumns if not Started')
        for(let i = 0; i < rawNames.length; i++){
            const rawName = rawNames[i]
            const name =  this.getIdentifier(rawName);
            this.status.columnInfo.push({name, ...rawColumninfo});
            this.status.columnProcessor.push(new TypeProcessor(new StringType(), name));
            this.columnInfering.push(this.typeConstructors().map(c =>new TypeInferer(createNew(c))));
        }
    }
    createTableSql(){
        var esto = this;
        switch(esto.status.is){
            case Status.Started:
                return "create table " + this.table_name + " (" + 
                    esto.status.columnInfo.map(info => "\n " + info.name + " " + info.type_name).join(",") + 
                "\n);\n";
            default: throw Error("wrong status at createTableSql: "+this.status);
        }
    }
    async processLine(line:string){
        switch(this.status.is){
            case Status.Started:
                if(this.status.part == Part.Head){
                    this.processColumns(line.split(this.options.field_separator));
                    await this.status.create_table_raw.write(this.createTableSql());
                    await this.status.create_table_raw.close();
                    this.status.part = Part.Body;
                }else{
                    const firstRow = !this.status.rowNumber;
                    const rawValues = line.split(this.options.field_separator);
                    if(this.options.infer_types){
                        rawValues.forEach((rawValue, columnNumber) => this.inferType(rawValue, columnNumber))
                    }
                    const dataLine = "('" + 
                    rawValues.map(value=>value.replace(/'/g, "''")).join("', '") + 
                    "')"
                    await this.status.inserts?.write(
                        (!this.options.multi_insert || firstRow ?
                            "insert into " + this.table_name + (
                                this.options.insert_columns ? " (" + 
                                this.status.columnInfo.map(i=>i.name).join(", ") + 
                                ") " 
                                : " "
                            ) + "values" + (this.options.multi_insert ? "\n " : " " ) : ",\n " 
                        ) + dataLine + (this.options.multi_insert ? "" : ";\n")
                    )
                    this.status.rowNumber++
                }
            break;
            default: throw Error("wrong status at process_end: "+this.status);
        }
    }
    async saveCreateTable(){
        switch(this.status.is){
            case Status.Started:
                var file = await this.owf.open('create_table')
                this.status.create_table = file
                for(var i = 0; i < this.status.columnProcessor.length; i++){
                    var columnInfers = this.columnInfering[i]
                    var columnInfer: TypeInferer<unknown, TypeBase<unknown>> | undefined
                    do{
                        columnInfer = columnInfers.pop()
                    } while (columnInfer != null && columnInfer.hasInvalid)
                    if(columnInfer != null && !columnInfer.hasInvalid){
                        var name = this.status.columnInfo[i].name
                        var typer = columnInfer.typer
                        this.status.columnProcessor[i] = new TypeProcessor(typer, name)
                    }
                }
                file.write(this.createTableSql())
                file.close()
            break;
            default: throw Error("wrong status at saveCreateTable: "+this.status)
        }
    }
    async processEnd(){
        switch(this.status.is){
            case Status.Started:
                await this.status.inserts.close();
                if(this.options.infer_types){
                    await this.saveCreateTable()
                }
                if(this.options.multi_insert){
                    await this.status.inserts?.write(";\n")
                }
                await this.owf.end();
                this.status = {is: Status.Done}
            break;
            default: throw Error("wrong status at process_end: "+this.status)
        }
    }
    async processSmallFile(fileName:string){
        const fh = await fs.open(fileName)
        const content = await fh.readFile('utf-8');
        await this.processStart();
        for(const line of content.split(/\r?\n/)){
            // eslint-disable-next-line no-await-in-loop
            await this.processLine(line);
        }
        await this.processEnd();
    }
    inferedInfo(){
        return this.columnInfering;
    }
}

if(process.argv.length>2){
    const reExt = /\.\w+$/;
    const fileName = process.argv[2];
    if(!fileName.match(reExt)){
        throw new Error("filename must has an extension")
    }
    const txt2sql = new Txt2Sql(new FileWritterFactory());
    txt2sql.processSmallFile(fileName);
}

File: rust/txt2sql/src/txt2sql.rs
use std::fmt::{Debug, Formatter, Result};
use std::fs::File;
use std::io::{prelude::*, Write, BufReader};
use std::path::{Path, PathBuf};
use regex::Regex;
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Txt2SqlOptions {
    #[serde(default)]
    field_separator: String,
    #[serde(default)]
    table_name: String
}

impl Default for Txt2SqlOptions {
    fn default() -> Self { Self { 
        field_separator: "|".to_string(), 
        table_name: "table1".to_string()
    } }
}

// fn txt2sql_default_field_separator() -> &'static str { "|"}

pub trait OutWritter {
    fn write(&mut self, paragraph: &str) -> ();
    fn close(&mut self) -> (); // no more writes/closes allowed
}

pub trait OutWritterFactory {
    fn open(&mut self, file_name: &str) -> Box<dyn OutWritter>;
    fn end(&mut self) -> (); // no more writes/closes allowed in any opened file
}

struct FileWritter {
    file: Option<File>
}

impl OutWritter for FileWritter {
    fn write(&mut self, paragraph: &str){
        match self.file {
            Some(ref mut opened_file) => { opened_file.write(paragraph.as_bytes()).expect("cannot write"); }
            None => panic!("file is no longer opened")
        }
    }
    fn close(&mut self){
        self.file = None;
    }
}

struct FileWritterFactory<'a>{
    extension: &'a str,
    base_dir: &'a str
}

impl<'a> OutWritterFactory for FileWritterFactory<'a> {
    fn open(&mut self, file_name: &str) -> Box<dyn OutWritter> {
        let mut path: PathBuf = [self.base_dir, file_name].iter().collect();
        if path.extension() == None {
            path = path.with_extension(self.extension);
        }
        let f = File::create(&path).unwrap_or_else( |_| panic!("Unable to create file '{}'", path.display()) );
        Box::new(FileWritter{
            file: Some(f)
        })
    }
    fn end(&mut self){}
}

enum Txt2SqlPart { Body, Head }

struct Txt2SqlOpenedMembers {
    columns: Vec<String>,
    inserts: Box<dyn OutWritter>,
    create_table: Box<dyn OutWritter>,
    part: Txt2SqlPart
}

impl Debug for Txt2SqlOpenedMembers {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("Txt2SqlOpenedMembers").finish()
    }
}

impl Debug for dyn OutWritter {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("OutWritter").finish()
    }
}

// Parece innecesario porque ya lo hice para OutWritter
impl Debug for FileWritter {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        f.debug_struct("FileWritter").finish()
    }
}

#[derive(Debug)]
enum Txt2SqlStatus { New, Started(Txt2SqlOpenedMembers), Done }

pub struct Txt2Sql<'a>{
    status: Txt2SqlStatus,
    owf: &'a mut dyn OutWritterFactory,
    options: Txt2SqlOptions
}

impl<'a> Txt2Sql<'a> {
    pub fn init(owf: &'a mut dyn OutWritterFactory, options:Txt2SqlOptions) -> Self{
        Self{
            status: Txt2SqlStatus::New,
            options,
            owf
        }
    }
    pub fn process_start(&mut self){
        match self.status {
            Txt2SqlStatus::New => {
                self.status = Txt2SqlStatus::Started(Txt2SqlOpenedMembers{
                    columns: Vec::new(),
                    create_table: self.owf.open("create_table"),
                    inserts: self.owf.open("inserts"),
                    part: Txt2SqlPart::Head                  
                });
            }
            _ => panic!("wrong status at process_start: {:#?}", self.status)
        }
    }
    pub fn process_line(&mut self, line: &str){
        match &mut self.status {
            Txt2SqlStatus::Started(this) => {
                match this.part {
                    Txt2SqlPart::Head => {
                        let re = Regex::new(r"\W").unwrap();
                        this.columns = line.split(self.options.field_separator.as_str()).map(
                            |column| re.replace_all(column, "_").to_lowercase()
                        ).collect();
                        this.create_table.write(
                            format!(
                                "create table {} (\n{}\n);\n", 
                                self.options.table_name,
                                this.columns.iter().map(
                                    |column| format!(" {} text", column)
                                ).collect::<Vec<String>>().join(",\n")
                            ).as_str()
                        );
                        this.create_table.close();
                        this.part = Txt2SqlPart::Body;
                    }
                    Txt2SqlPart::Body => {
                        let reg_exp_quote_values = Regex::new(r"'").unwrap();
                        this.inserts.write(
                            format!(
                                "insert into {} ({}) values ({});\n", 
                                self.options.table_name,
                                this.columns.join(", "),
                                line.split(self.options.field_separator.as_str()).map(
                                    |value| format!("'{}'", reg_exp_quote_values.replace_all(value, "''").into_owned())
                                ).collect::<Vec<String>>().join(", ")
                            ).as_str()
                        );
                    }
                }
            }
            _ => panic!("wrong status at process_line: {:#?}", self.status)
        }
    }
    pub fn process_end(&mut self){
        match &mut self.status {
            Txt2SqlStatus::Started(this) => {
                this.inserts.close();
                self.owf.end();
                self.status = Txt2SqlStatus::Done;
            }
            _ => panic!("wrong status at process_end: {:#?}", self.status)
        }
    }
}

// from https://stackoverflow.com/questions/30801031/read-a-file-and-get-an-array-of-strings
fn lines_from_file(file_name: impl AsRef<Path>) -> Vec<String> {
    let file = File::open(file_name).expect("no such file");
    let buf = BufReader::new(file);
    buf.lines()
        .map(|l| l.expect("Could not parse line"))
        .collect()
}

pub fn process_small_file(file_name: &str, base_dir: &str){
    let mut fwf = FileWritterFactory{ extension: "sql", base_dir: base_dir };
    let lines = lines_from_file(file_name);
    let mut t2s = Txt2Sql::init(&mut fwf, Txt2SqlOptions::default());
    t2s.process_start();
    for line in lines {
        t2s.process_line(&line);
    }
    t2s.process_end();
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::vecwritter::*;
        
    #[test]
    fn it_works() {
        let mut vwf = VecWritterFactory::init();

        let mut txt2sql = Txt2Sql::init(&mut vwf, Txt2SqlOptions::default());
        txt2sql.process_start();
        txt2sql.process_line("column1|Column 2");
        txt2sql.process_line("data11|data12");
        txt2sql.process_line("data21|O'Donnell");
        txt2sql.process_end();
        let mut keys = vwf.vecs.keys().collect::<Vec<&String>>();
        keys.sort_unstable();
        assert_eq!(keys, ["create_table", "inserts"]);
        assert_eq!(vwf.vecs.get("create_table").expect("create_table.sql").borrow().to_vec(), [
            "create table table1 (\n column1 text,\n column_2 text\n);\n"
        ]);
        assert_eq!(vwf.vecs.get("inserts").expect("inserts.sql").borrow().to_vec(), [
            "insert into table1 (column1, column_2) values ('data11', 'data12');\n",
            "insert into table1 (column1, column_2) values ('data21', 'O''Donnell');\n"
        ]);
    }
}
